src/OptimalFilter.jl

```diff
@@ -0,0 +1,382 @@
+
+
+end
+
+# Calculate the mean for the optimal proposal of the height
+function calculate_mean_height!(mean::AbstractArray{T,3}, height::AbstractArray{T,3},
+                                offline_matrices::OfflineMatrices, observations::AbstractVector{T},
+                                stations, model_params, filter_params) where T
+
+    # The arguments for the WΛWH decompositions are matrices that only have nonzero values
+    # at the indices of the stations. Store them as sparse arrays to save space.
+    mu21 = sparse(stations.ist, stations.jst,
+                  offline_matrices.R22_inv * (offline_matrices.mu20 * observations),
+                  model_params.nx+1, model_params.ny+1)
+    mu22 = model_params.obs_noise_std^(-2) * sparse(stations.ist, stations.jst,
+                                          observations,
+                                          model_params.nx+1, model_params.ny+1)
+
+    # Compute WΛWH decompositions, results are dense matrices, store them in buffers
+    # These correspond to mu21 and mu22 in Alex's code
+    WΛWH_decomposition!(offline_matrices.buf1, mu21, offline_matrices, model_params)
+    WΛWH_decomposition!(offline_matrices.buf2, mu22, offline_matrices, model_params)
+
+    # Compute the difference of the decomposition results, store in offline_matrices.buf1
+    # This corresponds to mu2 in Alex's code.
+    # TODO: Check if WΛWH is linear and you could swap the order
+    offline_matrices.buf2 .-= offline_matrices.buf1
+
+    mu10 = Vector{T}(undef, model_params.nobs)
+
+    # Loop over particles
+    for iprt = 1:filter_params.nprt
+
+        mul!(mu10, offline_matrices.R22_inv, get_values_at_stations(@view(height[:,:,iprt]), stations))
+        mu10_sparse = sparse(stations.ist, stations.jst, mu10, model_params.nx+1, model_params.ny+1)
+
+        # Compute decomposition of height values at stations times the inverse covariance matrix
+        # The argument corresponds to mu10 and the outcome to mu11 in Alex's code
+        WΛWH_decomposition!(offline_matrices.buf1, mu10_sparse, offline_matrices, model_params)
+
+        # Compute the mean for the ith particle using mu2 and mu11
+        # Skip storing the temporary mu1 in Alex's code
+        mean[:,:,iprt] .= @view(height[:,:,iprt]) .- offline_matrices.buf1 .+ offline_matrices.buf2
+
+    end
+
+end
+
+function sample_height_proposal!(height::AbstractArray{T,3},
+                                 offline_matrices::OfflineMatrices, online_matrices::OnlineMatrices,
+                                 observations::AbstractVector{T}, stations, model_params, filter_params,
+                                 rng::Random.AbstractRNG) where T
+
+    @assert filter_params.nprt % 2 == 0 "Number of particles must be even"
+
+    calculate_mean_height!(online_matrices.mean, height, offline_matrices, observations, stations, model_params, filter_params)
+
+    #TODO: Can you create a Normal(0,1) distribution of complex type?
+    nd = Normal(0,1)
```

Comment:
In any case we don't really need to define this, `randn` has the same properties